train.py

```diff
@@ -456,32 +456,35 @@ def eval(
             metric_tracker_val.save()
 
             ################################################################################ Saving models
-            if args.save:
-                state = {
-                    "args": args,
-                    "epoch": epoch,
-                    "model": model.state_dict(),
-                    "optimizer": optimizer.state_dict(),
-                    "scheduler": scheduler.state_dict(),
-                }
-
-                current_epoch_filename = f"{epoch}_ckpt.pth.tar"
-                latest_epoch_filename = "latest_ckpt.pth.tar"
 
-                save_checkpoint(
-                    state=state,
-                    directory=saved_models_filepath,
-                    filename=current_epoch_filename,
-                    is_best=False,
-                )
+            state = {
+                "args": args,
+                "epoch": epoch,
+                "model": model.state_dict(),
+                "optimizer": optimizer.state_dict(),
+                "scheduler": scheduler.state_dict(),
+            }
 
+            n_best_epochs = metric_tracker_val.get_best_n_epochs_for_metric(
+                "accuracy", n=args.save_top_n_val_models
+            )
+
+            if epoch in n_best_epochs:
```

Comment:
Do we really need to keep the 'rank' as part of the naming of the top N models? I can see why it's useful, but it we ommited that, we could just save a model if it's in the top 5, and simply delete the previous worst model. That would save on disk writes. What do you think? It shouldn't matter as it happens once per epoch, though. Having the literal rank in the name makes it a lot easier after the fact to find the model in a folder. hmmm..
What if we have a smarter version of what you are doing? Like, only store the new model that is in the top K, and rename the others to reflect the right rank? This must be a lot cheaper than saving them all again?

Reply:
You could literally keep a dict of ranks mapped to paths, and if the ranks change simply rename those that remain, and remove the one that doesn't, only saving the new one.

Reply:
> we could just save a model if it's in the top 5, and simply delete the previous worst model. That would save on disk writes

I'm not sure I understand. Wouldn't we be writing the same number of models either way? 

Unless I did it wrong, this should only save the current epoch weights, we should never re-save old things (+ we don't actually have access to the old weights, so we don't have anything to save there anyway)

I don't love the idea of having code which renames old checkpoints.

Reply:
Correct me if I am wrong, but the code seems to iterate the top n epoch idx and then save those models with a specific rank built into their names? If not, how do you handle the fact that the previous top model is now the second in rank?

Reply:
Ah you're right.

I guess we can't do this without having some renaming going on. I'll clear that up.